azurelinuxagent/common/utils/extensionprocessutil.py

```diff
@@ -68,12 +68,13 @@ def handle_process_completion(process, command, timeout, stdout, stderr, error_c
     process_output = read_output(stdout, stderr)
 
     if timed_out:
-        raise ExtensionError("Timeout({0}): {1}\n{2}".format(timeout, command, process_output),
-                             code=ExtensionErrorCodes.PluginHandlerScriptTimedout)
+        raise LaunchCommandError("Timeout({0}): {1}\n{2}".format(timeout, command, process_output),
```

Comment:
What is the intention of the new exception? How is it different to raise LaunchCommandError(code=ExtensionErrorCodes.PluginHandlerScriptTimedout) here vs ExtensionError(code=ExtensionErrorCodes.PluginHandlerScriptTimedout) @ SystemdCgroupsApi.start_command.

Also, when there is a timeout the exit code has no meaning.

If the intention was to capture the exit code of the extension then probably you could continue using ExtensionError(code=ExtensionErrorCodes.PluginHandlerScriptTimedout) for timeouts and rename the new exception to something like ExtensionCommanError and use it only for actual extension errors (which report an exit code).

Speaking of timeouts, it seems like we are raising timeout exceptions in multiple places. It may be better to raise the timeout from wait_for_process_completion_or_timeout (I would then rename it to wait_for_process_completion). That would match the design of Python's Subprocess.wait).

Reply:
> What is the intention of the new exception? How is it different to raise LaunchCommandError(code=ExtensionErrorCodes.PluginHandlerScriptTimedout) here vs ExtensionError(code=ExtensionErrorCodes.PluginHandlerScriptTimedout) @ SystemdCgroupsApi.start_command?

Yes the main reason for this exception was to differentiate between cases where the command execution failed with a non-0 exit code or if it failed due to some platform issue (unable to download handler manifest, etc).

> If the intention was to capture the exit code of the extension then probably you could continue using ExtensionError(code=ExtensionErrorCodes.PluginHandlerScriptTimedout) for timeouts and rename the new exception to something like ExtensionCommanError and use it only for actual extension errors (which report an exit code).

I agree, I'll modify the code accordingly. Also @vrdmr pointed out that we can reuse the `ExtensionOperationError` for this purpose as its pretty much being used for that. I'll see if I can reuse that error.

> It may be better to raise the timeout from wait_for_process_completion_or_timeout 

I totally agree to this, I was kind of under the impression we were already doing this. Let me refactor the code accordingly and modify the PR

Reply:
`ExtensionOperationError` and `ExtensionError` are not the same. `ExtensionOperationError` indicates a failure to launch the extension and `ExtensionError` indicates an error during extension execution.

Reply:
Sorry for the confusion, I meant `ExtensionOperationError` and `LaunchCommandError` look like they're there for similar purpose. I can add an extra property in the ExtensionOperationError if it doesn't break anything else.

Reply:
Also, the reason why we don't raise from `wait_for_process_completion_or_timeout` is because that would be too early in the systemd-run case (we want to determine if it was systemd that failed or the extension before raising `ExtensionError`). Check out `cgroupapi.py` starting from line 500.

Reply:
Just to keep in mind if you are to refactor this code. If you decide to raise from `wait_for_process_completion_or_timeout`, make sure to handle it appropriately in `cgroupapi.py#start_extension_command`.